File: 3493.properties-graph.java
#
# @lc app=leetcode id=3493 lang=java
#
# [3493] Properties Graph
#
# @lc code=start
class Solution {
    public int numberOfComponents(int[][] properties, int k) {
        int n = properties.length;
        List<Integer>[] adj = new List[n];
        for (int i = 0; i < n; i++) adj[i] = new ArrayList<>();
        // Build adjacency list
        for (int i = 0; i < n; i++) {
            Set<Integer> setI = new HashSet<>();
            for (int val : properties[i]) setI.add(val);
            for (int j = i + 1; j < n; j++) {
                Set<Integer> setJ = new HashSet<>();
                for (int val : properties[j]) setJ.add(val);
                Set<Integer> intersection = new HashSet<>(setI);
                intersection.retainAll(setJ);
                if (intersection.size() >= k) {
                    adj[i].add(j);
                    adj[j].add(i);
                }
            }
        }
        boolean[] visited = new boolean[n];
        int components = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                components++;
                dfs(i, adj, visited);
            }
        }
        return components;
    }
    private void dfs(int node, List<Integer>[] adj, boolean[] visited) {
        visited[node] = true;
        for (int nei : adj[node]) {
            if (!visited[nei]) dfs(nei, adj, visited);
        }
    }
}
# @lc code=end

File: 3493.properties-graph.go
#
# @lc app=leetcode id=3493 lang=golang
#
# [3493] Properties Graph
#
# @lc code=start
func numberOfComponents(properties [][]int, k int) int {
    n := len(properties)
    // Step 1: Define intersect function
    intersect := func(a, b []int) int {
        setA := make(map[int]struct{})
        for _, val := range a {
            setA[val] = struct{}{}
        }
        count := 0
        for _, val := range b {
            if _, found := setA[val]; found {
                count++
            }
        }
        return count
    }
    // Step 2: Build adjacency list based on intersect criteria (>= k)
    adjList := make([][]int, n)
    for i := 0; i < n; i++ {
        for j := i + 1; j < n; j++ {
            if intersect(properties[i], properties[j]) >= k {
                adjList[i] = append(adjList[i], j)
                adjList[j] = append(adjList[j], i) // undirected graph 
            } 
        } 
    } 
    // Step 3: Use DFS/BFS to find connected components & count them & return result & boolean array visited initialized as false & componentCount initialized as zero & inside dfs function in all adjacent nodes if not visited then call dfs function again with that node as argument & increment componentCount by one for each node that is not visited at first and return componentCount